README.md
# Purslane
<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->
[![All Contributors](https://img.shields.io/badge/all_contributors-1-orange.svg?style=flat-square)](#contributors-)
<!-- ALL-CONTRIBUTORS-BADGE:END -->

Scalable, flexible, affordable, and safe way to perform periodic tasks. All from a convenient CLI binary!

*Note:* 
1. Purslane is being used in production for 1 or 2 projects to test it out but that's the extent of the testing thus far. 
2. The project is young so breaking changes may occur. However, this project will do it's best to always be safe to use by following semantic versioning. 
3. Purslane has only been tested through use and is not unit tested. It is planned, but has not yet been completed. 

![logo](misc/logo.jpg)
> credits: PaoloBis / Getty Images

## Why use Purslane? 

Let's use an example. You have a production database that you want to perform daily backups on. This seems like a simple task, but there are some issues that you could run into:

* **What machine should I run my backup on?...** Performing database backups takes resources (CPU, memory) and you don't want a daily backup to slow down your servers so it's best to run backups on a separate machine with resources that can handle the backup. 
* **Disk storage...** As the size of your database grows, you will need larger disk storage to store the dump on. 
* **Pain to scale...** After you setup your server for database dumps, your app gets popular. As time goes on you will need to spend time upgrading your backup system to handle the new load.
* **Higher costs...** If I run backups daily, that means that 23 1/2 hours of each day that server is doing nothing and you still need to pay for it.

Purslane solves all of these problems for you! Let's read about how Purslane works to understand how all of these problems are gone.

## How does Purslane work?

Purslane is quite simple, really. Each time Purslane is run, it will:
1. Create a new cloud server and optionally attach disk storage to it. (Purslane works with [DigitalOcean](https://www.digitalocean.com/) at this time but may include other cloud providers in the future)
2. Run a Docker container of your choice in the newly created cloud server. 
3. When the Docker container exits, Purslane will destroy all of the resources it created such as the created server and disk storage. 

For you, this means:
* **Your application has no performance hit...** Because your task runs on a brand new server, your infrastructure does not take a hit. 
* **Scale with 1 line of code...** As the size of your application grows and you need to scale your periodic tasks, it only takes 1 quick change to your Purslane config file and Purslane will scale the created server and disk storage the next time you run your task. 
* **Affordable...** You only pay for the time your Docker container runs. 
* **Flexible...** Run any task you need! You are the only that makes the Docker container, so you can run anything you need to run! Any language, any software, any task. 

All you need to do is give Purslane authentication details to login to your cloud provider to create the server, set the size of the server and optional disk storage you want, and provide the Docker image to run. Purslane will take care of the rest. 

# Getting started

* The first thing you will need is to create a Docker image for Purslane to execute. You can host this Docker image in a public or private Docker image registry. 

* Create the Purslane config file that tells Purslane how to execute your Docker image. You can create it in the default location of `~/.purslane` or create it anywhere and point to it with the `--config` CLI argument. 

```yaml
cloud: # Required. Must set a cloud provider. At this time, Purslane only works with DigitalOcean
  digitalocean:
    # Create an API key with write access as we are creating resources. 
    # https://www.digitalocean.com/docs/apis-clis/api/create-personal-access-token/
    api_key: "123" 

volume: # Optional, but required if you want to create a volume that attaches to the server. 
  gigs: 1 # How many GBs you want the volume to be. 
  container_mount_path: "/home" # The path inside of your Docker container you want the volume to attach to. 

server: # Optional - sets default values for you if left out. 
  # this format is universal with all providers but some providers may not have the combination you specify. This format will be converted to the string specific to the cloud provider for you. 
  # the format of this string: `s-1cpu-1gb`
  # s - standard. Can also be "m" for high memory or "c" for dedicated CPU. 
  # 1cpu - how many CPUs you want. 
  # 1gb - how many gigs of memory you want. 
  size: "s-1cpu-1gb" 
  # private network to add server to when created. (optional)
  #  - DigitalOcean: Enter UUID of already created private network. Find UUID in DigitalOcean's website > select a VPC > See UUID in URL of webpage. 
  private_network: "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX" 

# Required. These keys are used to run commands against the server once it's created. The machine you run the CLI from will SSH into the created server and send commands to it. 
# **Note:** At this time, SSH keys must not have a passphrase on them 
public_ssh_key_path: "~/.ssh/id_rsa.pub"
private_ssh_key_path: "~/.ssh/id_rsa"

docker_run: # Optional. Purslane runs a default set that will run the docker container just fine. 
  extra_args: "-p 5000:5000" # Append arguments to the `docker run` command. Great place to add ports bindings or environment variables, for example. 

docker: # Required. 
  image: "image-name:tag" # Image with tag of docker image to pull. Can be from private or public repo. 
  registry: # optional. Only needed if image in private repo you need to authenticate with. 
    name: "name-of-registry"
    username: "docker-username"
    password: "docker-password-to-registry"
```

> Note: This file contains secret information such as API keys. It's recommended that you make this config file read-only by select users on your machine. 

* Time to execute Purslane. You decide how you would like to execute the CLI. [Run the CLI on a server](#run-cli-on-a-server) or [run via Docker](#run-cli-via-docker). 

### Run CLI on a server 

Since Purslane is a very low resources CLI application, you can run it directly on a server of yours. 

There are many ways to download the CLI. 
1. Download the binary directly from [GitHub releases](https://github.com/levibostian/purslane/releases). 
2. `curl -sf https://gobinaries.com/levibostian/Purslane | sh`
3. `brew install levibostian/tap/purslane`

You now have the CLI installed on your machine. Run `purslane --help` to learn about the CLI. 

### Run CLI via Docker

You can run the Purslane CLI via a Docker image. There is [a public Purslane Docker image](https://hub.docker.com/levibostian/purslane) for you to use. When you want to use the Docker image, you need to provide to the Docker image your config file, ssh public/private key files, and maybe you need to provide other files. 

Here is an example `docker run` statement to run the Docker purslane CLI:
```
docker run --rm -v $(pwd)/purslane.yaml:/config.yaml -v /home/.ssh/:/ssh levibostian/purslane:latest run --config /config.yaml
```

> Note: The home directory of the Docker image is set to `/root`. You can use `~/` in the config file to refer to this path. 

[The Dockerfile](Dockerfile) is setup quite easily. It just runs the purslane CLI and everything that you add to the end of the `docker run` command gets passed as arguments to the CLI. So the example above with `docker run ... run --config /config.yaml` means that you are running `/purslane run --config /config.yaml` inside of the Docker container. 

## Development 

Purslane is a Go lang program. To start developing Purslane is as simple as (1) cloning the repo and (2) running `go run main.go`.

## Contribute

Purslane is open for pull requests. Check out the [list of issues](https://github.com/levibostian/purslane/issues) for tasks planned to be worked on. Check them out if you wish to contribute in that way.

**Want to add features to Purslane?** Before you decide to take a bunch of time and add functionality to the CLI, please, [create an issue](https://github.com/levibostian/Purslane/issues/new) stating what you wish to add. This might save you some time in case your purpose does not fit well in the use cases of Purslane.

## Contributors 

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key))

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/levibostian"><img src="https://avatars1.githubusercontent.com/u/2041082?v=4" width="100px;" alt=""/><br /><sub><b>Levi Bostian</b></sub></a><br /><a href="https://github.com/levibostian/Purslane/commits?author=levibostian" title="Code">💻</a> <a href="https://github.com/levibostian/Purslane/commits?author=levibostian" title="Documentation">📖</a> <a href="#maintenance-levibostian" title="Maintenance">🚧</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->


CHANGELOG.md
# [1.5.0-beta.1](https://github.com/levibostian/purslane/compare/1.4.2...1.5.0-beta.1) (2020-12-28)


### Features

* add alpine docker image ([3f0f53f](https://github.com/levibostian/purslane/commit/3f0f53f0ca0366524676cc94d12386a8a6167d52))

## [1.4.2](https://github.com/levibostian/Purslane/compare/1.4.1...1.4.2) (2020-07-06)


### Bug Fixes

* link to github over twitter ([8a099e9](https://github.com/levibostian/Purslane/commit/8a099e95507b8198a78e6c1a40645fdfa6e799a5))

## [1.4.1](https://github.com/levibostian/Purslane/compare/1.4.0...1.4.1) (2020-07-05)


### Bug Fixes

* snapcraft config plugs fix ([479beb7](https://github.com/levibostian/Purslane/commit/479beb7beac29456fdc18f6b8d9351d8eb2ba230))

# [1.4.0](https://github.com/levibostian/Purslane/compare/1.3.0...1.4.0) (2020-07-05)


### Features

* configure snapcraft build/deploy ([b859354](https://github.com/levibostian/Purslane/commit/b8593549e172bd7264e4c626b025ffa1f9da5ef7))

# [1.3.0](https://github.com/levibostian/Purslane/compare/1.2.0...1.3.0) (2020-07-05)


### Features

* add homebrew deploy for purslane ([3b970d5](https://github.com/levibostian/Purslane/commit/3b970d52bbdd69490877983eb91e6836eaa86528))

# [1.2.0](https://github.com/levibostian/Purslane/compare/1.1.0...1.2.0) (2020-07-05)


### Features

* add dockerfile to run CLI in ([90381d0](https://github.com/levibostian/Purslane/commit/90381d0831b96778095f33a5e1c84be7ad09c6c6))

# [1.1.0](https://github.com/levibostian/Purslane/compare/1.0.1...1.1.0) (2020-07-04)


### Features

* add private networking ability ([00a4c19](https://github.com/levibostian/Purslane/commit/00a4c19149bb992a9f71c3ff699c15e3ba702f79))

## [1.0.1](https://github.com/levibostian/Purslane/compare/1.0.0...1.0.1) (2020-06-04)


### Bug Fixes

* cli help docs have non-default values ([1b2e388](https://github.com/levibostian/Purslane/commit/1b2e388a0688621a70379220b91e834115bc5b1c))

# 1.0.0 (2020-05-23)


### Bug Fixes

* server gets created and runs image success! ([9a8a866](https://github.com/levibostian/Purslane/commit/9a8a866ae53de125002abca9bb2caaaca0a836c6))


### Features

* bash scrpt created as proof of concept ([e2ba71d](https://github.com/levibostian/Purslane/commit/e2ba71d68942b2a9f70349b1d429d6e2b23492d3))
* cleanup resources after run docker image ([0cd3087](https://github.com/levibostian/Purslane/commit/0cd3087826b65ce9705bb683713d1349e9d6656a))
* create server ([c1c365c](https://github.com/levibostian/Purslane/commit/c1c365cd370bd2b62416eea5a26718d73e813eec))
* docker run command inside of server ([ae21b39](https://github.com/levibostian/Purslane/commit/ae21b396eb0e473e634e905ac4ffb32daa97b05d))
* prepare created server for running docker ([c97dffe](https://github.com/levibostian/Purslane/commit/c97dffe5d69620689bba009b85ca08f626dafeb7))
* setup CLI program to run commands ([d1d4754](https://github.com/levibostian/Purslane/commit/d1d4754fa4bf91d825197390d5007038eed6e091))


script.sh
#!/bin/bash

#----------------
# Required environment variables that you need to pass in
# DO_TOKEN - digitalocean api token
# DO_SSH_FINGERPRINT - fingerprint of ssh key registered to DO account to use. Go to https://cloud.digitalocean.com/account/security to add a key and to find the fingerprint. 
#----------------

# Notes found during dev to be helpful later. 
# * DO names for volumes and stuff requires names must be lowercase and alphanumeric. hyphens are ok. 
# * We have to use a ssh key to login to the droplet.
# * Can we use a user script to set environment variables on the system? if so, we sould try to set the docker login password so it's ready. 

# Create block storage volume 
echo "Creating volume"
CREATE_VOLUME_BODY='
{"size_gigabytes":1, "name": "purslane-testing-volume", "description": "Volume for project Purslane. Ok to destroy. Its just for testing.", "region": "nyc1", "filesystem_type": "ext4", "filesystem_label": "purslane"}
'
#curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer $DO_TOKEN" -d "$CREATE_VOLUME_BODY" "https://api.digitalocean.com/v2/volumes"
# Example response: {"volume":{"id":"ce45863d-8b50-11ea-ad6b-0a58ac14477a","name":"purslane-testing-volume","created_at":"2020-05-01T02:09:36Z","description":"Volume for project Purslane. Ok to destroy. Its just for testing.","droplet_ids":null,"region":{"features":["private_networking","backups","ipv6","metadata","install_agent","storage","image_transfer"],"name":"New York 1","slug":"nyc1","sizes":["s-1vcpu-1gb","512mb","s-1vcpu-2gb","1gb","s-3vcpu-1gb","s-2vcpu-2gb","s-1vcpu-3gb","s-2vcpu-4gb","2gb","s-4vcpu-8gb","m-1vcpu-8gb","c-2","4gb","g-2vcpu-8gb","m-16gb","s-6vcpu-16gb","c-4","8gb","m-2vcpu-16gb","m3-2vcpu-16gb","g-4vcpu-16gb","m6-2vcpu-16gb","m-32gb","s-8vcpu-32gb","c-8","16gb","m-4vcpu-32gb","m3-4vcpu-32gb","g-8vcpu-32gb","s-12vcpu-48gb","m6-4vcpu-32gb","m-64gb","s-16vcpu-64gb","c-16","32gb","m-8vcpu-64gb","m3-8vcpu-64gb","g-16vcpu-64gb","s-20vcpu-96gb","48gb","m6-8vcpu-64gb","m-128gb","s-24vcpu-128gb","64gb","g-32vcpu-128gb","s-32vcpu-192gb","m-24vcpu-192gb"],"available":true},"size_gigabytes":1,"filesystem_type":"ext4","filesystem_label":"purslane","tags":null}}

# You need to get the returned ID from the response because that's the ID you put into the next request to create a droplet with that volume attached. 

# Create droplet 
echo "Create droplet"
CREATE_DROPLET_BODY='
{"name":"purslane-testing-droplet","region":"nyc1","size":"s-1vcpu-1gb","image":"docker-18-04","backups":false,"ipv6":true,"user_data":null,"private_networking":null, "volumes": ["dd633e6a-8bac-11ea-ad6b-0a58ac14477a"]
'
CREATE_DROPLET_BODY="$CREATE_DROPLET_BODY, \"ssh_keys\": [\"$DO_SSH_FINGERPRINT\"] }"

# curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer $DO_TOKEN" -d "$CREATE_DROPLET_BODY" "https://api.digitalocean.com/v2/droplets"
# Example response: {"droplet":{"id":190584016,"name":"purslane-testing-droplet","memory":1024,"vcpus":1,"disk":25,"locked":false,"status":"active","kernel":null,"created_at":"2020-05-01T02:16:52Z","features":["ipv6"],"backup_ids":[],"next_backup_window":null,"snapshot_ids":[],"image":{"id":50944795,"name":"Docker 5:19.03.1~3 on 18.04","distribution":"Ubuntu","slug":"docker-18-04","public":true,"regions":["nyc3","nyc1","sfo1","nyc2","ams2","sgp1","lon1","nyc3","ams3","fra1","tor1","sfo2","blr1","sfo3"],"created_at":"2019-08-15T16:01:20Z","min_disk_size":20,"type":"application","size_gigabytes":0.8,"description":"Docker 5:19.03.1~3 on 18.04 20190815","tags":[],"status":"available"},"volume_ids":["265b43b2-8b51-11ea-ad6b-0a58ac14477a"],"size":{"slug":"s-1vcpu-1gb","memory":1024,"vcpus":1,"disk":25,"transfer":1.0,"price_monthly":5.0,"price_hourly":0.00744,"regions":["ams2","ams3","blr1","fra1","lon1","nyc1","nyc2","nyc3","sfo1","sfo2","sfo3","sgp1","tor1"],"available":true},"size_slug":"s-1vcpu-1gb","networks":{"v4":[{"ip_address":"142.93.124.171","netmask":"255.255.240.0","gateway":"142.93.112.1","type":"public"}],"v6":[{"ip_address":"2604:a880:400:d1::a7b:1001","netmask":64,"gateway":"2604:a880:400:d1::1","type":"public"}]},"region":{"name":"New York 1","slug":"nyc1","features":["private_networking","backups","ipv6","metadata","install_agent","storage","image_transfer"],"available":true,"sizes":["s-1vcpu-1gb","512mb","s-1vcpu-2gb","1gb","s-3vcpu-1gb","s-2vcpu-2gb","s-1vcpu-3gb","s-2vcpu-4gb","2gb","s-4vcpu-8gb","m-1vcpu-8gb","c-2","4gb","g-2vcpu-8gb","m-16gb","s-6vcpu-16gb","c-4","8gb","m-2vcpu-16gb","m3-2vcpu-16gb","g-4vcpu-16gb","m6-2vcpu-16gb","m-32gb","s-8vcpu-32gb","c-8","16gb","m-4vcpu-32gb","m3-4vcpu-32gb","g-8vcpu-32gb","s-12vcpu-48gb","m6-4vcpu-32gb","m-64gb","s-16vcpu-64gb","c-16","32gb","m-8vcpu-64gb","m3-8vcpu-64gb","g-16vcpu-64gb","s-20vcpu-96gb","48gb","m6-8vcpu-64gb","m-128gb","s-24vcpu-128gb","64gb","g-32vcpu-128gb","s-32vcpu-192gb","m-24vcpu-192gb"]},"tags":[]}}

# You want to ping the endpoint below and check the "status" key until it goes to "active" from "new" which means it's ready to connect to. 
# curl -X GET -H "Content-Type: application/json" -H "Authorization: Bearer $DO_TOKEN" "https://api.digitalocean.com/v2/droplets/190634206"
# example response: {"droplet":{"id":190634206,"name":"purslane-testing-droplet","memory":1024,"vcpus":1,"disk":25,"locked":false,"status":"active","kernel":null,"created_at":"2020-05-01T13:10:08Z","features":["ipv6"],"backup_ids":[],"next_backup_window":null,"snapshot_ids":[],"image":{"id":50944795,"name":"Docker 5:19.03.1~3 on 18.04","distribution":"Ubuntu","slug":"docker-18-04","public":true,"regions":["nyc3","nyc1","sfo1","nyc2","ams2","sgp1","lon1","nyc3","ams3","fra1","tor1","sfo2","blr1","sfo3"],"created_at":"2019-08-15T16:01:20Z","min_disk_size":20,"type":"application","size_gigabytes":0.8,"description":"Docker 5:19.03.1~3 on 18.04 20190815","tags":[],"status":"available"},"volume_ids":["dd633e6a-8bac-11ea-ad6b-0a58ac14477a"],"size":{"slug":"s-1vcpu-1gb","memory":1024,"vcpus":1,"disk":25,"transfer":1.0,"price_monthly":5.0,"price_hourly":0.00744,"regions":["ams2","ams3","blr1","fra1","lon1","nyc1","nyc2","nyc3","sfo1","sfo2","sfo3","sgp1","tor1"],"available":true},"size_slug":"s-1vcpu-1gb","networks":{"v4":[{"ip_address":"178.128.149.127","netmask":"255.255.240.0","gateway":"178.128.144.1","type":"public"}],"v6":[{"ip_address":"2604:a880:400:d1::97a:3001","netmask":64,"gateway":"2604:a880:400:d1::1","type":"public"}]},"region":{"name":"New York 1","slug":"nyc1","features":["private_networking","backups","ipv6","metadata","install_agent","storage","image_transfer"],"available":true,"sizes":["s-1vcpu-1gb","512mb","s-1vcpu-2gb","1gb","s-3vcpu-1gb","s-2vcpu-2gb","s-1vcpu-3gb","s-2vcpu-4gb","2gb","s-4vcpu-8gb","m-1vcpu-8gb","c-2","4gb","g-2vcpu-8gb","m-16gb","s-6vcpu-16gb","c-4","8gb","m-2vcpu-16gb","m3-2vcpu-16gb","g-4vcpu-16gb","m6-2vcpu-16gb","m-32gb","s-8vcpu-32gb","c-8","16gb","m-4vcpu-32gb","m3-4vcpu-32gb","g-8vcpu-32gb","s-12vcpu-48gb","m6-4vcpu-32gb","m-64gb","s-16vcpu-64gb","c-16","32gb","m-8vcpu-64gb","m3-8vcpu-64gb","g-16vcpu-64gb","s-20vcpu-96gb","48gb","m6-8vcpu-64gb","m-128gb","s-24vcpu-128gb","64gb","g-32vcpu-128gb","s-32vcpu-192gb","m-24vcpu-192gb"]},"tags":[]}}

# You get the ip address from the response. and because you added your ssh key, you're ready to perform the ssh connection once the status is active. 
# ssh into the droplet with `ssh root@ip-address` but make sure you run ssh with flags to accept the fingerprint yes/no question. 
# login to registry `echo $DOCKER_LOGIN_PASSWORD | docker login docker.pkg.github.com --username levibostian --password-stdin` this works without a password prompt. make sure to allow the user to configure this line because they might use a differnet registry or might use a public one. 

# Run docker container `docker run --rm -v /home/app/extra_stuff:/mnt/purslane_testing_volume/ docker.pkg.github.com/levibostian/purslane/test-docker-image:latest`
#   I will want to add more stuff to this like environment variables and port forwarding so the container can talk to a DB via ssh tunnel. this is somethign i still need to test. my docker container used for testing was too basic and we didn't test ssh tunnels and port binding. 
#   I will want to capture the exit code and stdout/stderr of the docker container to map that stderr/stdout to the CLI's stdout/stderr and make sure the exit code of the CLI is the exit code of the docker container running. 

# Time to delete the droplet and it's resources. This endpoint used below takes an array of volumes, snapshots, etc to delete along with droplet. There is another endpoint where you can delete everything attached to the droplet but we will use this one instead as it's safer in case we decide to attach to other stuff in the future. Only delete the volume we created for you. 
echo "Deleting resources"

# Path of the url is the droplet ID. the body contains the volume ID created above. 
# curl -X DELETE -H "Content-Type: application/json" -H "Authorization: Bearer $DO_TOKEN" -d '{"volumes": ["dd633e6a-8bac-11ea-ad6b-0a58ac14477a"]}' "https://api.digitalocean.com/v2/droplets/190634206/destroy_with_associated_resources/selective"



